Java/core/src/main/java/com/amazon/randomcutforest/tree/RandomCutTree.java

```diff
@@ -300,7 +312,7 @@ public Integer addPoint(Integer pointIndex, long sequenceIndex) {
                     if (currentBox.contains(point) || parent == Null) {
```

Comment:
The process of adding a node is bottom up; with the caveat that any cut at the higher level (closer to root) invalidates previous cuts (e.g., look at parentPath.clear() inside the while(true) loop). Interestingly this makes the Update Tree operation the same in spirit as the traversal. It is possible to write a traversal that updates the tree in this manner -- but that may become more opaque.

Reply:
Maybe we can do the following change since another randomCut is not needed:

```
if (savedDim == Integer.MAX_VALUE) {
                        randomCut(factor, point, currentBox);
                        throw new IllegalStateException(" cut failed ");
                    }
```

can be changed to 

```
if (savedDim == Integer.MAX_VALUE) {
         throw new IllegalStateException(" cut failed ");
}
```

Reply:
For the other part - you're right, we can probably use the new checkArgument to raise the exception; and producing a new random cut is not required